File: src/trello.ts
const root = "https://api.trello.com";

interface Board {
  id: string;
  desc: string;
  descData: string;
  closed: boolean;
  idOrganization: string;
  pinned: boolean;
  url: string;
  shortUrl: string;
  prefs: {
    permissionLevel: string;
    hideVotes: boolean;
    voting: "disabled";
    comments: string;
    selfJoin: boolean;
    cardCovers: boolean;
    isTemplate: boolean;
    cardAging: "pirate";
    calendarFeedEnabled: boolean;
    background: string;
    backgroundImage: string;
    backgroundImageScaled: string;
  };
  labelNames: {
    green: string;
    yellow: string;
    orange: string;
    red: string;
    purple: string;
    blue: string;
    sky: string;
    lime: string;
    pink: string;
    black: string;
  };
  limits: {
    attachments: {
      perBoard: {
        status: "ok";
        disableAt: number;
        warnAt: number;
      };
    };
  };
  starred: boolean;
  memberships: string;
  shortLink: string;
  subscribed: boolean;
  powerUps: string;
  dateLastActivity: string;
  dateLastView: string;
  idTags: string;
  datePluginDisable: string;
  creationMethod: string;
  ixUpdate: number;
  templateGallery: string;
  enterpriseOwned: boolean;
}

interface Card {
  id: string;
  address: string;
  badges: {
    attachmentsByType: {
      trello: {
        board: number;
        card: number;
      };
    };
    location: boolean;
    votes: number;
    viewingMemberVoted: boolean;
    subscribed: boolean;
    fogbugz: string;
    checkItems: number;
    checkItemsChecked: number;
    comments: number;
    attachments: number;
    description: boolean;
    due: string;
    dueComplete: boolean;
  };
  checkItemStates: [string];
  closed: boolean;
  coordinates: string;
  creationMethod: string;
  dateLastActivity: string;
  desc: string;
  descData: {
    emoji: {};
  };
  due: string;
  dueReminder: string;
  email: string;
  idBoard: string;
  idChecklists: string[];
  idLabels: string[];
  idList: string;
  idMembers: [string];
  idMembersVoted: [string];
  idShort: number;
  labels: {
    id: string;
    idBoard: string;
    name: string;
    color: string;
  }[];
  limits: {
    attachments: {
      perBoard: {
        status: string;
        disableAt: number;
        warnAt: number;
      };
    };
  };
  locationName: string;
  manualCoverAttachment: boolean;
  name: string;
  pos: number;
  shortLink: string;
  shortUrl: string;
  subscribed: boolean;
  url: string;
  cover: {
    color: string;
    idUploadedBackground: boolean;
    size: string;
    brightness: string;
    isTemplate: boolean;
  };
}

export interface ChecklistItem {
  idChecklist: string;
  state: "complete";
  idMember: null;
  id: string;
  name: string;
  nameData: null;
  pos: number;
  due: null;
}

export interface Checklist {
  id: string;
  name: string;
  idCard: string;
  pos: number;
  idBoard: string;
  checkItems: ChecklistItem[];
}

export type TrelloClient = ReturnType<typeof trello>;
export interface RequestFactory {
  (
    method: "get" | "delete" | "post" | "put",
    url: string,
    params: {},
    data?: unknown
  ): Promise<{}>;
}

export const trello = (opts: {
  key: string;
  token: string;
  makeRequest: RequestFactory;
}) => {
  const { key, token, makeRequest } = opts;

  return {
    getBoard: (id: string) =>
      makeRequest("get", `${root}/1/boards/${id}`, { key, token }) as Promise<
        Board
      >,
    getBoardChecklists: (id: string) =>
      makeRequest("get", `${root}/1/boards/${id}/checklists`, {
        key,
        token,
        fields: "all",
      }) as Promise<Board>,
    getBoardCards: (id: string) =>
      makeRequest("get", `${root}/1/boards/${id}/cards`, {
        key,
        token,
      }) as Promise<Card[]>,
    getCard: (id: string) =>
      makeRequest("get", `${root}/1/cards/${id}`, { key, token }) as Promise<
        Card
      >,

    getChecklist: (id: string) =>
      makeRequest("get", `${root}/1/checklists/${id}`, {
        key,
        token,
      }) as Promise<Checklist>,
    updateChecklist: (id: string, data: Checklist) =>
      makeRequest("put", `${root}/1/checklists/${id}`, { key, token }, data),

    updateChecklistItem: (
      idCard: string,
      idChecklist: string,
      idCheckItem: string,
      data: ChecklistItem
    ) =>
      makeRequest(
        "put",
        `${root}/1/cards/${idCard}/checklist/${idChecklist}/checkItem/${idCheckItem}`,
        { key, token },
        data
      ),

    // Add a checklist item to a checklist
    addChecklistItem: (
      idChecklist: string,
      name: string,
      pos?: number
    ) =>
      makeRequest(
        "post",
        `${root}/1/checklists/${idChecklist}/checkItems`,
        { key, token },
        { name, pos }
      ) as Promise<ChecklistItem>,

    // Get cards in a specific list
    getListCards: (idList: string) =>
      makeRequest("get", `${root}/1/lists/${idList}/cards`, {
        key,
        token,
      }) as Promise<Card[]>,

    // Move a card to a different list
    moveCardToList: (idCard: string, idList: string) =>
      makeRequest(
        "put",
        `${root}/1/cards/${idCard}`,
        { key, token },
        { idList }
      ),

    removeLabel: (idCard: string, idLabel: string) =>
      makeRequest("delete", `${root}/1/cards/${idCard}/idLabels/${idLabel}`, {
        key,
        token,
      }),

    // Get all lists on a board
    getBoardLists: (idBoard: string) =>
      makeRequest("get", `${root}/1/boards/${idBoard}/lists`, {
        key,
        token,
      }),

    // Create a new checklist on a card
    createChecklist: (idCard: string, name: string = "Checklist") =>
      makeRequest(
        "post",
        `${root}/1/cards/${idCard}/checklists`,
        { key, token },
        { name }
      ) as Promise<Checklist>,
  };
};

File: shopr/trello.py
"""Trello API client for shopr."""

import logging
from typing import Any

import httpx
from pydantic import BaseModel, ConfigDict


logger = logging.getLogger("shopr:trello")

ROOT = "https://api.trello.com"


class Card(BaseModel):
    """Trello card representation.

    Only includes fields actually used by the application.
    Unknown fields from the API are ignored.
    """

    model_config = ConfigDict(extra="ignore")

    id: str
    name: str
    idChecklists: list[str] = []
    labels: list[dict[str, Any]] = []


class ChecklistItem(BaseModel):
    """Trello checklist item representation.

    Only includes fields actually used by the application.
    Unknown fields from the API are ignored.
    """

    model_config = ConfigDict(extra="ignore")

    id: str
    idChecklist: str
    name: str
    pos: int | float = 0
    state: str = "incomplete"


class Checklist(BaseModel):
    """Trello checklist representation.

    Only includes fields actually used by the application.
    Unknown fields from the API are ignored.
    """

    model_config = ConfigDict(extra="ignore")

    id: str
    idCard: str = ""
    checkItems: list[ChecklistItem] = []


class TrelloClient:
    """Client for interacting with the Trello API."""

    def __init__(self, key: str, token: str):
        """Initialize the Trello client.

        Args:
            key: Trello API key
            token: Trello API token
        """
        self.key = key
        self.token = token

    async def _request(
        self,
        method: str,
        url: str,
        params: dict[str, Any] | None = None,
        data: Any = None,
    ) -> Any:
        """Make an HTTP request."""
        all_params = {"key": self.key, "token": self.token, **(params or {})}
        async with httpx.AsyncClient() as client:
            response = await client.request(
                method=method.upper(),
                url=url,
                params=all_params,
                json=data,
            )
            response.raise_for_status()
            return response.json()

    async def get_board_checklists(self, id: str) -> list[Checklist]:
        """Get all checklists on a board."""
        result = await self._request(
            "get", f"{ROOT}/1/boards/{id}/checklists", {"fields": "all"}
        )
        return [Checklist.model_validate(item) for item in result]

    async def get_board_cards(self, id: str) -> list[Card]:
        """Get all cards on a board."""
        result = await self._request("get", f"{ROOT}/1/boards/{id}/cards")
        return [Card.model_validate(item) for item in result]

    async def get_card(self, id: str) -> Card:
        """Get a card by ID."""
        result = await self._request("get", f"{ROOT}/1/cards/{id}")
        return Card.model_validate(result)

    async def get_checklist(self, id: str) -> Checklist:
        """Get a checklist by ID."""
        result = await self._request("get", f"{ROOT}/1/checklists/{id}")
        return Checklist.model_validate(result)

    async def update_checklist(self, id: str, data: Checklist) -> dict[str, Any]:
        """Update a checklist."""
        return await self._request(
            "put", f"{ROOT}/1/checklists/{id}", data=data.model_dump()
        )

    async def update_checklist_item(
        self,
        id_card: str,
        id_checklist: str,
        id_check_item: str,
        data: ChecklistItem,
    ) -> dict[str, Any]:
        """Update a checklist item."""
        return await self._request(
            "put",
            f"{ROOT}/1/cards/{id_card}/checklist/{id_checklist}/checkItem/{id_check_item}",
            data={"name": data.name, "pos": data.pos, "state": data.state},
        )

    async def add_checklist_item(
        self,
        id_checklist: str,
        name: str,
        pos: int | None = None,
    ) -> ChecklistItem:
        """Add a checklist item to a checklist."""
        data: dict[str, Any] = {"name": name}
        if pos is not None:
            data["pos"] = pos
        result = await self._request(
            "post", f"{ROOT}/1/checklists/{id_checklist}/checkItems", data=data
        )
        return ChecklistItem.model_validate(result)

    async def get_list_cards(self, id_list: str) -> list[Card]:
        """Get cards in a specific list."""
        result = await self._request("get", f"{ROOT}/1/lists/{id_list}/cards")
        return [Card.model_validate(item) for item in result]

    async def move_card_to_list(self, id_card: str, id_list: str) -> dict[str, Any]:
        """Move a card to a different list."""
        return await self._request(
            "put", f"{ROOT}/1/cards/{id_card}", data={"idList": id_list}
        )

    async def remove_label(self, id_card: str, id_label: str) -> dict[str, Any]:
        """Remove a label from a card."""
        return await self._request(
            "delete", f"{ROOT}/1/cards/{id_card}/idLabels/{id_label}"
        )

    async def get_board_lists(self, id_board: str) -> list[dict[str, Any]]:
        """Get all lists on a board."""
        return await self._request("get", f"{ROOT}/1/boards/{id_board}/lists")

    async def create_checklist(
        self,
        id_card: str,
        name: str = "Checklist",
    ) -> Checklist:
        """Create a new checklist on a card."""
        result = await self._request(
            "post", f"{ROOT}/1/cards/{id_card}/checklists", data={"name": name}
        )
        return Checklist.model_validate(result)
